src/challenge.rs

```diff
@@ -0,0 +1,58 @@
+use ark_ff::PrimeField;
+use ark_sponge::FieldBasedCryptographicSponge;
+
+/// State stored for univariate generator
+#[derive(Copy, Clone)]
+pub struct UnivariateGeneratorState<F: PrimeField> {
+    gen: F,
+    next: F,
+}
+
+impl<F: PrimeField> UnivariateGeneratorState<F> {
+    fn new(gen: F) -> Self {
+        Self { gen, next: gen }
+    }
+
+    fn get_next(&mut self) -> F {
+        let result = self.next;
+        self.next *= self.gen;
+        result
+    }
+}
```

Comment:
Does this need to be an independent struct? Can we just move this into `ChallengeGenerator::Univariate`?

Reply:
I see that it makes the code later cleaner. Let's keep it as is.

Reply:
Yes! Then, we can have something like `ChallengeGenerator::Univariate(F, F)`